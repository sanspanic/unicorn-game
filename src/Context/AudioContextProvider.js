import React, { useRef } from "react";
import AudioContext from "./AudioContext";
import useSound from "use-sound";
import Bump from "../Assets/Audio/bump.wav";
import Move from "../Assets/Audio/move.wav";
import Lose from "../Assets/Audio/lose.wav";
import Win from "../Assets/Audio/win.wav";

const AudioContextProvider = ({ children }) => {
  //plays on illegal move
  const [playIllegal] = useSound(Bump, { volume: 0.25 });

  //plays on legal move
  const [playLegal] = useSound(Move, { volume: 0.25 });

  //plays on game over
  const [playGameOver] = useSound(Lose, { volume: 0.25 });

  //plays on game won
  const [playGameWon] = useSound(Win, { volume: 0.25 });

  return (
    <AudioContext.Provider
      value={{ playIllegal, playLegal, playGameOver, playGameWon }}
    >
      {children}
    </AudioContext.Provider>
  );
};

export default AudioContextProvider;
