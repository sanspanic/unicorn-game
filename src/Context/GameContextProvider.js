import React, { useState } from "react";
import GameContext from "./GameContext";

const GameContextProvider = ({ children }) => {
  const initialState = {
    status: "won",
    difficulty: 0,
    size: 15,
    name: "",
  };
  const [gameData, setGameData] = useState(initialState);

  //restarts game by setting gameData.status back to "setup" & redirecting to homepage
  const playAgain = () => {
    setGameData({ ...gameData, status: "setup" });
  };

  return (
    <GameContext.Provider value={{ gameData, setGameData, playAgain }}>
      {children}
    </GameContext.Provider>
  );
};

export default GameContextProvider;
