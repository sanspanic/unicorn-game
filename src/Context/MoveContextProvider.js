import React, { useState, useContext, useEffect } from "react";
import MoveContext from "./MoveContext";
import UnicornApi from "../API/UnicornApi";
import GameContext from "./GameContext";
import { getSpritePositions } from "../Game/Maze/mazeHelpers";

const MoveContextProvider = ({ children }) => {
  const [spritePositions, setSpritePositions] = useState({});
  const { gameData, setGameData } = useContext(GameContext);
  const [mazeId, setMazeId] = useState("");

  //sets focus on maze grid to enable hotkeys
  //unfortunately useRef didn't work so I'm using, inappropriately, useEffect
  useEffect(() => {
    if (mazeId === "") return;
    console.log("SELECTING DOM ELEMENT TO FOCUS, mazeId: ", mazeId);
    const Maze = document.querySelector("[tabindex='-1']");
    Maze.focus();
  }, [mazeId]);

  //handles move
  const handleMove = (direction) => {
    const move = async (data) => {
      try {
        const res = await UnicornApi.makeMove(data, mazeId);
        console.log("MADE A MOVE: ", res);
        //if (res.state === "won") setGameState("won");
        if (res["state-result"] === "Move accepted") {
          updatePositions();
        } else {
          //TODO: handle illegal move
        }
        checkForGameOver(res);
      } catch (e) {
        console.log(e);
      }
    };
    move({ direction });
  };

  //updates positions of sprites following a move
  const updatePositions = async () => {
    try {
      const res = await UnicornApi.getMazeCurrentState(mazeId);
      setSpritePositions(getSpritePositions(res));
      console.log("UPDATED SPRITE POSITIONS AFTER MOVE: ", spritePositions);
    } catch (e) {
      console.log(e);
    }
  };

  //checks if game is won or lost
  const checkForGameOver = (res) => {
    if (res.state === "won") {
      setGameData({ ...gameData, status: "won" });
    } else if (res.state === "over") {
      setGameData({ ...gameData, status: "over" });
    }
  };

  return (
    <MoveContext.Provider
      value={{
        handleMove,
        updatePositions,
        checkForGameOver,
        mazeId,
        setMazeId,
        spritePositions,
        setSpritePositions,
      }}
    >
      {children}
    </MoveContext.Provider>
  );
};

export default MoveContextProvider;
