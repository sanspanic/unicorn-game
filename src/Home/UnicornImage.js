import React from "react";
import Unicorn from "../Assets/Imgs/Unicorn.png";
import ReactFloaterJs from "react-floaterjs";

const UnicornImage = () => {
  return (
    <ReactFloaterJs>
      <img className="w-5/12 sm:w-3/12" src={Unicorn}></img>
    </ReactFloaterJs>
  );
};

export default UnicornImage;
