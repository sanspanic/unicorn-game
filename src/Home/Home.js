import React, { useContext } from "react";
import { Link } from "react-router-dom";
import UnicornImage from "./UnicornImage";

const Home = () => {
  return (
    <div className="p-5 h-screen ">
      <h1 className="font-black text-pink-600 text-4xl py-5 sm:py-10">
        Hi. I am a Unicorn.{" "}
      </h1>
      <div className="flex flex-col font-mono justify-evenly h-5/6">
        <UnicornImage />
        <p>
          No, I'm not a privately held startup company valued at over $1
          billion.
        </p>
        <p>
          I'm the real deal, a mystical creature of magic as well as Scotland's{" "}
          <a
            className="underline"
            href="https://www.visitscotland.com/about/uniquely-scottish/national-animal-unicorn/"
          >
            {" "}
            national animal
          </a>
          .
        </p>
        <p>
          I'm being hunted by a Monster. Can I{" "}
          <span className="text-pink-600 font-semibold">TRUST</span> you to{" "}
          <span className="text-pink-600 font-semibold">PILOT</span> me safely
          to the rainbow?
        </p>
        <button className="gradient-pink border-2 border-purple-600 text-white px-3 py-3 shadow-xl rounded mr-auto">
          {" "}
          <Link to="/game"> Yes! Let's go!</Link>
        </button>
      </div>
    </div>
  );
};

export default Home;
