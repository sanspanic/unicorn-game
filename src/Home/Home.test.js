import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Home from "./Home";
import { BrowserRouter, MemoryRouter } from "react-router-dom";
import { createMemoryHistory } from "history";
import GameContextProvider from "../Context/GameContextProvider";
import MoveContextProvider from "../Context/MoveContextProvider";
import AudioContextProvider from "../Context/AudioContextProvider";

test("renders without crashing", () => {
  render(
    <MemoryRouter>
      <Home />
    </MemoryRouter>
  );
});

test("matches snapshot", () => {
  const { asFragment } = render(
    <MemoryRouter>
      <Home />
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});
