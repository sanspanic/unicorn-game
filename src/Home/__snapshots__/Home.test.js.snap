// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="p-5 h-screen "
  >
    <h1
      class="font-black text-pink-600 text-4xl py-5 sm:py-10"
    >
      Hi. I am a Unicorn. 
    </h1>
    <div
      class="flex flex-col font-mono justify-evenly h-5/6"
    >
      <div
        class="react-floater-animated"
      >
        <img
          alt="floating unicorn"
          class="w-5/12 sm:w-3/12"
          src="Unicorn.png"
        />
      </div>
      <p>
        No, I'm not a privately held startup company valued at over $1 billion.
      </p>
      <p>
        I'm the real deal, a mystical creature of magic as well as Scotland's 
        <a
          class="underline"
          href="https://www.visitscotland.com/about/uniquely-scottish/national-animal-unicorn/"
        >
           national animal
        </a>
        .
      </p>
      <p>
        I'm being hunted by a Monster. Can I 
        <span
          class="text-pink-600 font-semibold"
        >
          TRUST
        </span>
         you to 
        <span
          class="text-pink-600 font-semibold"
        >
          PILOT
        </span>
         me safely to the rainbow?
      </p>
      <button
        class="gradient-pink border-2 border-purple-600 text-white px-3 py-3 shadow-xl rounded mr-auto"
      >
         
        <a
          href="/game"
        >
           Yes! Let's go!
        </a>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
